Add tab stops support for wxListBox (windows-only).

artur-so opened the pull request:
Add wxLB_USETABSTOPS so that '\t' (tab) characters in wxListBox strings get recognized and expanded. The flag forwards to the native LB_USETABSTOPS one.
Add wxListBox::SetTabStops() to optionally specify tab stop distances. It is equivalent to using native LB_SETTABSTOPS message.
Illustrate the usage of tab stops in widgets/listbox sample.

artur-so committed:
Add tab stops support for wxListBox (windows-only).

Add wxLB_USETABSTOPS so that '\t' (tab) characters in wxListBox strings get recognized and expanded. The flag forwards to the native LB_USETABSTOPS one.
Add wxListBox::SetTabStops() to optionally specify tab stop distances. It is equivalent to using native LB_SETTABSTOPS message.
Illustrate the usage of tab stops in widgets/listbox sample.

artur-so commented:
Some notes:
- `wxListBox` docs mention: _Notice that currently TAB characters in list box items text are not handled consistently under all platforms, so they should be replaced by spaces to display strings properly everywhere. The list box doesn't support any other control characters at all._
- I've checked on macOS and linux+gtk3 and tabs are supported by default, although I've seen no way of setting tab stops. So maybe the docs paragraph above needs to be revised?
- Maybe wxLB_USETABSTOPS should be enabled by default to achieve similar behavior across the three major platforms?
- As I've seen `wxComboBox` behaves similarly, that is, it supports tabs by default on mac and linux. But there is no way to use tabs in the native `COMBOBOX` control on msw.

vadz commented:
Thanks, this looks fine as it is, but I agree that we could turn `LB_USETABSTOPS` on by default if this makes MSW more consistent with the other platforms because I don't really see how could anything be broken by this -- i.e. I just can't imagine any existing programs relying on TABs being just stripped from strings, which seems to be the current behaviour.

I also wonder if we shouldn't rename `SetTabStops()` to `MSWSetTabStops()` if there is no way to implement it under the other platforms anyhow (we use `MSW` (or `GTK` etc) prefixes for the functions to make it obvious that they're only available in the given port).

In any case, the additions to the API need to be documented: this is the only part which is missing, so could you please update `interface/wx/listbox.h` accordingly?

TIA!

artur-so commented:
Thanks for the review!

> I agree that we could turn LB_USETABSTOPS on by default if this makes MSW more consistent with the other platforms.

Ok, I'll do it. I'll also remove the `wxLB_USETABSTOPS` MSW-only flag. I think the "TAB"- related note should also be removed from docs then.
> I also wonder if we shouldn't rename SetTabStops() to MSWSetTabStops()

I also wondered, but acted according to the context. The method just above `SetHorizontalExtent()` is also MSW-specific and has no prefix. I haven't checked whether it is possible to set tab stops on other platforms. Please let me know how you still prefer regarding the prefix.

> the additions to the API need to be documented

No problem, will do when the above is clarified.

vadz commented:
> > I agree that we could turn LB_USETABSTOPS on by default if this makes MSW more consistent with the other platforms.
> 
> Ok, I'll do it. I'll also remove the `wxLB_USETABSTOPS` MSW-only flag. 

Yes, I agree, we don't even need a flag here as only one behaviour is actually useful.

> I think the "TAB"- related note should also be removed from docs then.

This not is in the same `interface/wx/listbox.h` file, so please feel free to remove or update it in this PR. TIA!

> > I also wonder if we shouldn't rename SetTabStops() to MSWSetTabStops()
> 
> I also wondered, but acted according to the context. 

Ah, I see. Sorry, this context wasn't immediately visible when looking at the patch.

> The method just above `SetHorizontalExtent()` is also MSW-specific and has no prefix. 

This method exists since at least 1998 (start of VCS history) and probably much earlier, so it predates the use of these prefixes by at least a decade or so :-/

> I haven't checked whether it is possible to set tab stops on other platforms. Please let me know how you still prefer regarding the prefix.

After a quick look, I don't see any way to customize this neither for GTK nor for Cocoa, so let's use the prefix. If we find a way (and the need) to implement this for the other platforms later, we can always add a prefix-less function and keep the old one for compatibility, it won't cost anything.

> > the additions to the API need to be documented
> 
> No problem, will do when the above is clarified.

TIA once again!

artur-so commented:
Just checked `interface/wx/listbox.h` and found no other MSW-specific method is present.
So I want to make sure that `MSWSetTabStops()` is well-suited there?
PS: the TAB-related remark I have updated locally.

vadz commented:
> Just checked `interface/wx/listbox.h` and found no other MSW-specific method is present.
> So I want to make sure that `MSWSetTabStops()` is well-suited there?

There are some documented MSW-only methods, a quick grep finds `MSWGetTaskBarButton()`, `MSWUseToasts()`, `MSWGetSystemMenu()` and `MSWSetEmulationLevel()`, so this wouldn't be the only exception if we do document it.

Maybe it would be more clear to put it the other way round: I don't know why should we have it if it's not documented, it's not like we need it internally. IOW, both of the following are acceptable:

- Have this method and document it.
- Not have it at all.

But I'd like to avoid adding it without documentation.

artur-so committed:
Add tab stops support for wxListBox (windows-only) - follow-up.

Always set the LB_USETABSTOPS style flag to achieve compatible behaviour with other platforms (linux/GTK and macOS/Cocoa).
Remove wxLB_USETABSTOPS as it does not make sense to ever disable it thus making the tabs characters ignored.
Add MSW prefix to wxListBox::SetTabStops method name. It is not implemented on other platforms.
Update wxListBox documentation, i.e. remove remark that TABs are not supported and document MSWSetTabStops().

artur-so commented:
Added the follow-up changes as discussed.
I assume the `@since` tag in documentation is added with first release it is included in.

vadz commented:
Looks good, thanks! I'd prefer to use `wxVector<int>` instead of `wxArrayInt`, but it's not a big deal.

I'll merge this once the CI builds pass, thanks again!

PBfordev commented on interface/wx/listbox.h:
```diff
@@ -321,6 +318,20 @@ class wxListBox : public wxControl,
+        This function is not available in the other ports by design, any
+        occurrences of it in the portable code must be guarded by
+        @code #ifdef __WXMSW__ @endcode preprocessor guards.
+
```
Shouldn't the `@since 3.1.4` tag be here?

Not sure if the instructions about guarding MSW-specific code are really necessary...

artur-so replied:
> Shouldn't the @since 3.1.4 tag be here?
I also wondered about the @since tag in a comment above, not knowing who/when adds it.
> Not sure if the instructions about guarding MSW-specific code are really necessary...
I got inspired (copy-pasted) from docs of other MSW-specific methods. Will gladly remove it, just please confirm.

PBfordev commented on src/msw/listbox.cpp:
```diff
@@ -611,6 +614,11 @@ void wxListBox::SetHorizontalExtent(const wxString& s)
 }
 
+void wxListBox::MSWSetTabStops(const wxArrayInt& tabStops)
+{
```
I wonder if adding something like
` wxCHECK_RET(!tabStops.empty(), "tabStops can't be empty");` 
to avoid dereferencing possibly invalid `tabStops.begin()` would be of any use...

And a supernitpick, but I would prefer using `size()` instead of` Count()`...

artur-so replied:
I'll apply both you suggestions and will also change `wxArrayInt` to `wxVector<int>` as @vadz mentioned.
Thank you for the review!

vadz commented:
Just FYI, I'll wait for the updates you mentioned before merging this then.

artur-so committed:
Follow-up PR review. Minor changes.

artur-so commented:
I have checked in the discussed changes, except for:
`wxCHECK_RET(!tabStops.empty(), "tabStops can't be empty");`
@PBfordev as you also mentioned empty tab stops array have a special reset-to-default meaning.
So I just added a safety check and passed NULL in such case.

PBfordev commented on interface/wx/listbox.h:
```diff
@@ -321,6 +318,22 @@ class wxListBox : public wxControl,
+
+        @onlyfor{wxmsw}
+
+        @since 3.1.0
```
```suggestion
        @since 3.1.4
```

artur-so committed:
Update interface/wx/listbox.h

Change version tag from @since 3.1.0 to @since 3.1.4

Co-Authored-By: PB <PBforDev@gmail.com>

artur-so commented:
@PBfordev Only now I noticed this comment of yours:
> Is there a reason why the function does not return a bool to mimic the native API?
No reason. I'll change this, too.
PS: did you edit your comment? I guess there are no notifications sent in such case?

PBfordev commented:
> PS: did you edit your comment? I guess there are no notifications sent in such case?
I have not edited the comment since the day I posted, 13 days ago, as can be seen in its edit history.

The only thing I have done recently here was suggesting the `@since` change today.

artur-so commented:
> I have not edited that comment 

Ok, then I missed the return value part, sorry for that.

To my defense, I checked how this is handled in the same `listbox.cpp` file and didn't see a consistent approach. For most set/action messages the return value is not checked.
In any case, I'm also in favor of a safer approach, so I'll add this.

artur-so committed:
Change return type of wxListBox::MSWSetTabStops() to bool.

vadz commented:
Thanks, I'm going to apply this (with just some minor changes to the doc/comments) in a moment.
